worker/file_util.py

```diff
@@ -0,0 +1,187 @@
+                       exclude_patterns=[], exclude_names=[]):
+    """
+    Returns a file-like object containing a tarred and gzipped archive of the
+    given directory.
+
+    follow_symlinks: Whether symbolic links should be followed.
+    ignore_names: Any top-level directory entries with names in ignore_names
+                  are not included.
+    exclude_patterns: Any directory entries with the given names at any depth in
+                      the directory structure are excluded.
+    """
+    args = ['tar', 'czf', '-', '-C', directory_path]
+    if follow_symlinks:
+        args.append('-h')
+    if exclude_patterns:
+        for pattern in exclude_patterns:
+            args.append('--exclude=' + pattern)
+    names = [name for name in os.listdir(directory_path) if name not in exclude_names]
+    if names:
+        args.extend(names)
+    else:
+        args.extend(['--files-from', '/dev/null'])
+    try:
+        proc = subprocess.Popen(args, stdout=subprocess.PIPE)
+        return proc.stdout
+    except subprocess.CalledProcessError as e:
+        raise IOError(e.output)
+
+
+def un_tar_gzip_directory(fileobj, directory_path):
+    """
+    Extracts the given file-like object containing a tarred and gzipped archive
+    into the given directory. The directory should already exist.
+
+    Raises tarfile.TarError if the archive is not valid.
+    """
+    directory_path = os.path.realpath(directory_path)
+    with tarfile.open(fileobj=fileobj, mode='r|gz') as tar:
+        for member in tar:
+            # Make sure that there is no trickery going on (see note in
+            # TarFile.extractall() documentation.
+            member_path = os.path.realpath(os.path.join(directory_path, member.name))
+            if not member_path.startswith(directory_path):
+                raise tarfile.TarError('Archive member extracts outside the directory.')
+
+            tar.extract(member, directory_path)
+
+
+def gzip_file(file_path, follow_symlinks=False):
+    """
+    Returns a file-like object containing the gzipped version of the given file.
+    """
+    if follow_symlinks:
+        file_path = os.path.realpath(file_path)
+    elif os.path.islink(file_path):
+        raise IOError('Not following symbolic links.')
+    args = ['gzip', '-c', '-n', file_path]
+    try:
+        proc = subprocess.Popen(args, stdout=subprocess.PIPE)
+        return proc.stdout
+    except subprocess.CalledProcessError as e:
+        raise IOError(e.output)
+
+
+def un_gzip_stream(fileobj):
+    """
+    Returns a file-like object containing the contents of the given file-like
+    object after gunzipping.
+
+    Raises an IOError if the archive is not valid.
+    """
+    class UnGzipStream(object):
+        def __init__(self, fileobj):
+            self._fileobj = fileobj
+            self._decoder = zlib.decompressobj(16 + zlib.MAX_WBITS)
+            self._buffer = ''
+            self._finished = False
+    
+        def read(self, num_bytes=None):
```

Comment:
Here's the comment from the code:

    # Note, that we don't use gzip.GzipFile or the gunzip shell command since
    # they require the input file-like object to support either tell() or
    # fileno(). Our version requires only read() and close().


Same for tar. Subprocess requires fileno() which is not available in the RPCFileHandle.